projects/extension/src/background/ConnectionManager.ts

```diff
@@ -194,25 +236,119 @@ export class ConnectionManager extends (EventEmitter as { new(): StateEmitter })
   async addChain(
     name: string,
     chainSpec: string,
-    jsonRpcCallback: SmoldotJsonRpcCallback): Promise<SmoldotChain> {
+    jsonRpcCallback: SmoldotJsonRpcCallback,
+    tabId?: number): Promise<Network> {
     if (!this.#client) {
       throw new Error('Smoldot client does not exist.');
     }
+    const existingNetwork = this.#networks.find(n => n.name === name && n.tabId === tabId)
+    if (existingNetwork)
+      return existingNetwork
 
     const addedChain = await this.#client.addChain({
       chainSpec,
       jsonRpcCallback,
       potentialRelayChains: this.#networks.map(net => net.chain),
     });
 
-    this.#networks.push({
+    const network: Network = {
+      tabId: tabId || 0,
       name,
       chain: addedChain,
       status: 'connected',
       isKnown: true,
       chainspecPath: `${name}.json`
-    });
+    }
+
+    this.#networks.push(network);
+
+    return network;
+  }
+
+  /** Handles the incoming message that contains Spec. */
+  #handleSpecMessage = (msg: MessageToManager, app: AppProps): void => {
```

Comment:
You've missed converting the test for this.  It should be easy to do- you want to check that the `app.chain` has been set after sending the spec message (similar setup to the 'App already disconnected' test but checking what happens after the spec message

Reply:
Actually I have not - in tests now I am using the actual ConnectionManager and not the ConnectionManagerMock. I will try to figure out a way to test these, if possible with the ConnectionManager or else I will need to adjust the ConnectionManagerMock. Im not sure though if these should be tested with a mock - these are actually covered from our sample apps (that we use for test)